File: luscii_patient_actions_sdk_android/android/src/main/kotlin/nl/digizorg/extensions/Action+Extension.kt
package nl.digizorg.extensions

import com.luscii.sdk.actions.Action

fun Action.toMap(): Map<String, Any?> {
    return mapOf(
        "id" to id.toString(),
        "name" to name,
        "icon" to icon?.toString(), // Convert URL to string
        "completedAt" to completedAt?.toEpochSecond(), // Convert ZonedDateTime to timestamp (seconds since epoch)
        "launchableStatus" to serializeLaunchableStatus(),
        "isLaunchable" to isLaunchable()
    )
}

private fun Action.serializeLaunchableStatus(): String {
    return when (val status = getLaunchableStatus()) {
        is Action.LaunchableStatus.Launchable -> {
            when (status) {
                is Action.LaunchableStatus.CompletedAt -> {
                    val epochSecond = status.date.toEpochSecond()
                    "completed:$epochSecond"
                }
                else -> "launchable"
            }
        }
        is Action.LaunchableStatus.NotLaunchable -> {
            when (status) {
                is Action.LaunchableStatus.LaunchableAfter -> {
                    val epochSecond = status.date.toEpochSecond()
                    "after:$epochSecond"
                }
                is Action.LaunchableStatus.LaunchableBefore -> {
                    val epochSecond = status.date.toEpochSecond()
                    "before:$epochSecond"
                }
                else -> "not_launchable"
            }
        }
        else -> "unknown"
    }
}

File: luscii_patient_actions_sdk_ios/ios/luscii_patient_actions_sdk_ios/Sources/luscii_patient_actions_sdk_ios/extensions/Action+Extension.swift
import Actions

@available(iOS 17.0, *)
extension Action {
  func toMap() -> [String: Any?] {
    return [
      "id": id.uuidString,
      "name": name,
      "icon": icon?.absoluteString,  // Convert URL to string
      "completedAt": completedAt?.timeIntervalSince1970,  // Convert Date to timestamp
      "launchableStatus": serializeLaunchableStatus(),
      "isLaunchable": isLaunchable,
      "isPlanned": isPlanned,
      "isSelfCare": isSelfCare,
      "isExtra": isExtra
    ]
  }
  
  private func serializeLaunchableStatus() -> String {
    switch launchableStatus {
    case .launchable:
      return "launchable"
    case .completed(let date):
      return "completed:\(date.timeIntervalSince1970)"
    case .after(let date):
      return "after:\(date.timeIntervalSince1970)"
    case .before(let date):
      return "before:\(date.timeIntervalSince1970)"
    @unknown default:
      return "unknown"
    }
  }
}
